nemo_aligner/models/nlp/gpt/gpt_sft_model.py

```diff
@@ -98,3 +133,31 @@ def finish_validation_step(self):
         mbs = int(self.cfg.data.train_ds.micro_batch_size)
         dp_size = int(parallel_state.get_data_parallel_world_size())
         configure_batch_sizes(mbs=mbs, gbs=gbs, dp=dp_size)
+
+    @torch.no_grad()
+    def infer(self, inference_batch, length_params=None, sampling_params=None, strategy=None):
+        prompt_tokens = inference_batch["text"].cuda(non_blocking=True)
+        prompt_lengths = inference_batch["length"].cuda(non_blocking=True)
+        inputs = (prompt_tokens, prompt_lengths)
+
+        default_params = self.get_inference_params()
+
+        full_length_params = default_params["length_params"].copy()
+        full_length_params.update(length_params or {})
+
+        full_sampling_params = default_params["sampling_params"].copy()
+        full_sampling_params.update(sampling_params or {})
+
+        inference_strategy = strategy or default_params["strategy"]
+
+        # need to disable activation checkpoint granularity for inference
+        self._reset_activation_checkpointing_args()
```

Comment:
This makes me realize that we should probably use an `Alignable` subclass and call `prepare_for_inference()`.

One issue is that inheriting from both `SupervisedInterface` and `AlignableInterface` is clunky. I ran into a similar situation in the past and IIRC I ended up creating a wrapper class (inheriting from `AlignableInterface`) on top of my model (though finally that code was dropped).

But there is actually a more fundamental issue: `SupervisedInterface` is meant to be compatible with our `SupervisedTrainer`, and this PR breaks this compatibility (by making `SupervisedTrainer` call `model.infer()`).

==> I feel like we should take a second look at our abstractions. @gshennvm would you happen to see a clear & clean solution to this?

Reply:
> I feel like we should take a second look at our abstractions.

agreed.

I think currently it is not clear to the user what each trainer expects, so they do not know what functions they need to implement for a trainer to support this and the abstractions are scattered. 

2 solutions I can think of:

1. We use a generic model configuration file and add in an abstraction say `AlignableModel` that covers **all** convenience functions(`{prepare,finish}_{training_step,validation_step,training,validation,inference}` and implement it for the GPTModel. This way when we have situations like the above, Igor can just call prepare_for_inference and it'll be correct.

The issue with the above is that our interface covers too much ground, someone with a new model that wants to only run say RM training will be forced to either inherit or implement something useless(like prepare_for_inference).

2. In each trainer will define a `REQUIRED_FUNCTIONS` attribute. And in the `__init__` of the trainer we check via `getattr` to make sure the model implemented all these required functions. If they didn't implement something like `infer` and run generation is turned on, then we should error. This lets us do away from forced interfaces and let's users confine to whatever trainer they want to build their model for.

Problem with this approach is that our trainers no longer share some structured API, and therefore will start drifting in what `REUQIRED_FUNCTIONS` attributes they expect, also there's not a great way of checking the function call signature of these `REQUIRED_FUNCTIONS` without complicated code.

Maybe as an intermediate solution we can do the following:
1. trainers all implement `REQUIRED_FUNCTIONS` to make it clear to the user what it needs
2. We will pass in a model with an abstraction(such as `AlignableInterface`) that implements a lot of the boilerplate(e.g `prepare_for_inference`)
3. in the trainer `__init__` we duck type the model to make sure it has the proper function attributes
4. MegatronGPTCritic, MegatronDPOModel, MegatronPPOActor, MegatronGPTRegressionRewardModel,MegatronGPTRewardModel should all inherit from a common class that implements `{prepare,finish}_{training_step,validation_step,training,validation,inference}` and implement `get_loss_and_metrics` and `infer` if applicable.

how does that sound?

Reply:
Discussed offline - these changes will go to a follow up PR

Reply:
Just to write down my thoughts on this:
* I would indeed go for a single model abstraction (`AlignableInterface`) to avoid class hierarchy hell.
* There are various ways we can address the problem of "how do we know what is required for a specific algorithm". This is something we can look into *after* the move to a single abstraction. Just to throw another option in the mix, we could have each algorithm implement a wrapper class that inherits from `AlignableInterface` but only forwards to the model the methods that are used by the algorithm (other methods would raise the default `NotImplementedError` from `AlignableInterface`). This way (1) this class can be used as documentation to show which methods are required, and (2) the algorithm is forced to use it so it can't cheat and access other methods without updating the class first. If we want to make dynamic requirements (that depend on config flags) explicit, then the wrapper class can be given access to the config and only forward the call when some flag is set.